File: src/contactdata/ContactData.kt
package contactdata

import app.*
import i18n.Translator
import react.*
import react.dom.*

external interface ContactData {
    val name: String
    val address: String
    val linkedin: String
    val github: String
    val bitbucket: String
    val phone: String?
    val email: String?
}

interface ContactDataState : RState {
    var data: ContactData?
}

class ContactDataComponent : RComponent<LanguageState, ContactDataState>() {

    override fun ContactDataState.init(props: LanguageState) {
        state.data = null
    }

    override fun componentWillReceiveProps(nextProps: LanguageState) {
        fetchDataFromJsonFile<ContactData>("data/contactdata/${nextProps.selectedLanguage}.json") {
            setState { data = it }
        }
    }

    override fun RBuilder.render() {
        contentSection(classes = "teal darken-3", id = "contactSection") {
            contentTitle(Translator.getTranslation("contactData"))
            state.data?.let {
                ul("collection") {
                    listItem(
                            iconGen = {
                                img(classes = "circle responsive-img", src = "images/profile_photo.jpg") {}
                            },
                            title = Translator.getTranslation("nameAndSurname"),
                            contentGen = { p { +it.name } }
                    )

                    simpleListItem("place", Translator.getTranslation("address"), {
                        p {
                            +it.address
                        }
                    }, "indigo")
                    it.email?.let { email ->
                        simpleListItem("contact_mail", Translator.getTranslation("email"), {
                            p {
                                +email
                            }
                        }, "purple darken-2")
                    }
                    it.phone?.let { phone ->
                        simpleListItem("contact_phone", Translator.getTranslation("phone"), {
                            p {
                                +phone
                            }
                        }, "amber darken-4")
                    }
                    simpleListItem("public", Translator.getTranslation("social"), {
                        div("row") {
                            div("col s4") {
                                a(href = it.linkedin, classes = "waves-effect waves-lime btn-flat") {
                                    i("small social-linkedin") {}

                                    span("no-text-transform hide-on-small-only") {
                                        +" LinkedIn"
                                    }
                                }
                            }
                            div("col s4") {
                                a(href = it.github, classes = "waves-effect waves-lime btn-flat") {
                                    i("small social-github") {}
                                    span("no-text-transform hide-on-small-only") {
                                        +" GitHub"
                                    }
                                }
                            }
                            div("col s4") {
                                a(href = it.bitbucket, classes = "waves-effect waves-lime btn-flat") {
                                    i("small social-bitbucket") {}
                                    span("no-text-transform hide-on-small-only") {
                                        +" Bitbucket"
                                    }
                                }
                            }
                        }
                    }, "lime darken-1")
                }
            }
        }
    }
}

fun RBuilder.contactData(selectedLanguage: String) {
    child(ContactDataComponent::class) {
        attrs.selectedLanguage = selectedLanguage
    }
}

private fun RBuilder.simpleListItem(icon: String, title: String, contentGen: RBuilder.() -> ReactElement, backgroundColorStyle: String = "blue") =
        listItem(
                iconGen = { i("material-icons circle darken-2 " + backgroundColorStyle) { +icon } },
                title = title,
                contentGen = contentGen,
                backgroundColorStyle = backgroundColorStyle
        )

private fun RBuilder.listItem(iconGen: RBuilder.() -> ReactElement, title: String, contentGen: RBuilder.() -> ReactElement, backgroundColorStyle: String = "blue") {
    li("collection-item avatar") {
        iconGen()
        span("title") { +title }
        contentGen()
    }
}

File: lib/sections/contact_data.dart
import 'package:cv_app_base/components/content_section.dart';
import 'package:cv_app_base/providers/contact_data.dart';
import 'package:cv_app_base/translation.dart';
import 'package:jaspr/html.dart';
import 'package:jaspr_riverpod/jaspr_riverpod.dart';

class ContactData extends StatelessComponent {
  @override
  Iterable<Component> build(BuildContext context) sync* {
    yield ContentSection(
      child: Builder(
        builder: (context) sync* {
          final contactData = context.watch(contactDataProvider);
          yield div(
            contactData.when(
              data: (data) => [
                img(src: 'public/images/profile_photo.jpg',
                        classes: [
                          'circle',
                          'responsive-img',
                        ],),
                        ul([
                          _ListItem(
                            icon: 'person',
                            title: span([Text(context.i18n('nameAndSurname'))]),
                            content: span([Text(data.name)]),
                          ),
                          _ListItem(
                            icon: 'place',
                            title: span([Text(context.i18n('address'))]),
                            content: span([Text(data.address)]),
                          ),
                          _ListItem(
                            icon: 'public',
                            title: span([Text(context.i18n('social'))]),
                            content: 
                            div([
                              _SocialRow(href: data.linkedin, name: 'LinkedIn', icon: 'social-linkedin'),
                              _SocialRow(href: data.github, name: 'GitHub', icon: 'social-github'),
                              _SocialRow(href: data.bitbucket, name: 'Bitbucket', icon: 'social-bitbucket'),
                            ])
                            ,
                          ),
                        ])
              ],
              error: (error, stackTrace) => [Text('Error: $error')],
              loading: () => [],
            ),
        );
        },
      ),
      id: 'contactData',
      title: context.i18n('contactData'),
    );
    
  }
}

class _ListItem extends StatelessComponent {
  final String icon;
  final Component title;
  final Component content;

  _ListItem({required this.icon, required this.title, required this.content});

  @override
  Iterable<Component> build(BuildContext context) sync* {
    yield li([
      span([Text(icon)], classes: ['material-icons', 'circle']),
      DomComponent(
        tag: 'dl',
        children: [
         DomComponent(tag: 'dt', child: title),
          DomComponent(tag: 'dd',child: content),
        ]),
    ]);
  }
}

class _SocialRow extends StatelessComponent {

  final String href;
  final String name;
  final String icon;

  _SocialRow({ required this.href, required this.name, required this.icon});

  @override
  Iterable<Component> build(BuildContext context) sync*{
    yield div([
      a([
        i([], classes: ['small', icon]),
        span([Text(name)], classes: ['no-text-transform', 'hide-on-small-only',]),
      ], href: href),
    ], classes: ['col', 's4']);
  }

}
